Architecture.md

```diff
@@ -0,0 +1,170 @@
+<!-- This file contains a rough overview of PyO3 codebase. -->
+<!-- Please do not make descriptions too specific, so that we can easily -->
+<!-- keep this file in sync with the codebase. -->
+
+# PyO3: Architecture.md
+
+This document roughly describes the high-level architecture of PyO3.
+If you are to become familiar with the codebase, you are in the right place!
+
+## Overview
+
+PyO3 provides a bridge between Rust and Python, based on the [Python C/API].
+Thus, PyO3 has low-level bindings of these API as its core.
+On top of that, we have higher-level bindings to operate Python objects safely.
+Also, to define Python class and functions in Rust code, we have `trait PyClass<T>` and a set of
+protocol traits (e.g., `PyIterProtocol`) for supporting object protocols (i.e., `__dunder__` methods).
+Since implementing `PyClass` requires lots of boilerplates, we have a proc-macro `#[pyclass]`.
+
+To summarize, we have mainly four parts in the PyO3 codebase.
+1. Low-level bindings of Python C/API.
+  - [`src/ffi`]
+2. Bindings to Python objects.
+  - [`src/instance.rs`], [`src/types`]
+3. `PyClass<T>` and related functionalities
+  - [`src/pycell.rs`], [`src/pyclass.rs`]
+4. Protocol methods like `__getitem__`.
+  - [`src/class`]
+5. Defining a Python class requires lots of glue codes, so we provide proc-macros to simplify the procedure.
+  - [`src/derive_utils.rs`]
+  - [`pyo3-macros`], [`pyo3-macros-backend`]
+
+## Low-level bindings of CPython API
+[`src/ffi`] contains wrappers of [Python C/API](https://docs.python.org/3/c-api/).
+
+We aim to provide straight-forward Rust wrappers and resemble the file structure of the
+[`cpython/Include`](https://github.com/python/cpython/tree/v3.9.2/Include).
+
+However, we still lack some API and continue to refactor the module to resemble the CPython's
+file structure completely.
+However, we still lack some API and continue to refactor the module to completely resemble
+the CPython's file structure.
+The tracking issue is [#1289](https://github.com/PyO3/pyo3/issues/1289), and contribution is welcome.
+
+## Bindings to Python Objects
+[`src/types`] contains bindings to [built-in types](https://docs.python.org/3/library/stdtypes.html)
+of Python, such as `dict` and `list`.
+Due to a historical reason, Python's `object` is called `PyAny` and placed in [`src/types/any.rs`].
+Currently, `PyAny` is a straight-forward wrapper of `ffi::PyObject`, like:
+```rust
+#[repr(transparent)]
+pub struct PyAny(UnsafeCell<ffi::PyObject>);
+```
+.
+
+All built-in types are defined as a C struct.
+For example, `dict` is defined as:
+```c
+typedef struct {
+    /* Base object */
+    PyObject ob_base;
+    /* Number of items in the dictionary */
+    Py_ssize_t ma_used;
+    /* Dictionary version */
+    uint64_t ma_version_tag;
+    PyDictKeysObject *ma_keys;
+    PyObject **ma_values;
+} PyDictObject;
+```
+.
+
+However, we cannot access such a specific data structure with `#[cfg(Py_LIMITED_API)]` set.
```

Comment:
maybe explain the feature flags at the top?

Reply:
👍 - it could be worth talking briefly about our `build.rs` also.